packages/expo-updates/ios/EXUpdates/EXUpdatesFileDownloader.m

```diff
@@ -0,0 +1,116 @@
+//  Copyright © 2018 650 Industries. All rights reserved.
+
+#import <EXUpdates/EXUpdatesFileDownloader.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+NSString * const kEXUpdatesFileDownloaderErrorDomain = @"EXUpdatesFileDownloader";
+NSTimeInterval const kEXUpdatesDefaultTimeoutInterval = 60;
+
+@interface EXUpdatesFileDownloader () <NSURLSessionDataDelegate>
+@end
+
+@implementation EXUpdatesFileDownloader
+
+- (void)downloadFileFromURL:(NSURL *)url
+                     toPath:(NSString *)destinationPath
+               successBlock:(EXUpdatesFileDownloaderSuccessBlock)successBlock
+                 errorBlock:(EXUpdatesFileDownloaderErrorBlock)errorBlock
+{
+  [self downloadDataFromURL:url successBlock:^(NSData * _Nonnull data, NSURLResponse * _Nonnull response) {
+    if ([[NSFileManager defaultManager] createFileAtPath:destinationPath contents:data attributes:nil]) {
+      successBlock(data, response);
+    } else {
+      errorBlock([NSError errorWithDomain:kEXUpdatesFileDownloaderErrorDomain code:-1 userInfo:@{NSLocalizedDescriptionKey: [NSString stringWithFormat:@"Could not write to path: %@", destinationPath]}], response);
+    }
+  } errorBlock:errorBlock];
+}
+
+- (void)downloadDataFromURL:(NSURL *)url
+               successBlock:(EXUpdatesFileDownloaderSuccessBlock)successBlock
+                 errorBlock:(EXUpdatesFileDownloaderErrorBlock)errorBlock
+{
+  NSURLSessionConfiguration *configuration = _urlSessionConfiguration ?: [NSURLSessionConfiguration defaultSessionConfiguration];
+
+  // also pass any custom cache policy onto this specific request
+  NSURLRequestCachePolicy cachePolicy = _urlSessionConfiguration ? _urlSessionConfiguration.requestCachePolicy : NSURLRequestUseProtocolCachePolicy;
+
+  NSURLSession *session = [NSURLSession sessionWithConfiguration:configuration delegate:self delegateQueue:nil];
```

Comment:
We should do more research on how NSURLSessions correspond to HTTP/2 streams and connections. Each connection can have multiple streams and we want to avoid needing to re-open a lot of new connections.

Reply:
i did some research and from what i understand, NSURLSession tasks will open a connection if one is not already open and will use one or more streams for a given task. however, NSURLSession does not allow fine-grained control over HTTP/2 connections, i.e. it will automatically close connections after they have been idle for around a minute and doesn't support keeping HTTP/2 connections open indefinitely. however, we can share a single session in each instance of FileDownloader (sessions are meant to coordinate a group of related tasks) which might reuse a connection if multiple tasks are sent before it's closed automatically. i've pushed a change to that effect.